File: lesson-3_dynamic-programming/0-1-knapsack-problem/recursion/main.cpp
#include <iostream>
#include <vector>
#include <fmt/core.h>

int getFav(int i, int balance, const std::vector<int> &values, const std::vector<int> &like);

int main() {
    int n, balance;
    std::cin >> n >> balance;

    auto values = std::vector<int>(n);
    auto like = std::vector<int>(n);
    for (int i = 0; i < n; i++) {
        std::cin >> values[i] >> like[i];
    }
    fmt::print("{}\n", getFav(n - 1, balance, values, like));
    return 0;
}

int getFav(int i, int balance, const std::vector<int> &values, const std::vector<int> &like) {
    if (i == -1 || balance <= 0) {
        return 0;
    }
    if (values[i] > balance) {
        return getFav(i - 1, balance, values, like);
    }
    return std::max(
            getFav(i - 1, balance, values, like),
            getFav(i - 1, balance - values[i], values, like) + like[i]
    );
}

File: lesson-3_dynamic-programming/0-1-knapsack-problem/recursion/main.go
package main

import (
    "fmt"
    "math"
)

func main() {
    var n, balance int
    fmt.Scan(&n, &balance)

    values := make([]int, n)
    like := make([]int, n)
    for i := 0; i < n; i++ {
        fmt.Scan(&values[i], &like[i])
    }
    fmt.Println(getFav(n-1, balance, values, like))
}

func getFav(i, balance int, values, like []int) int {
    if i == -1 || balance <= 0 {
        return 0
    }
    if values[i] > balance {
        return getFav(i-1, balance, values, like)
    }
    return int(math.Max(
        float64(getFav(i-1, balance, values, like)),
        float64(getFav(i-1, balance-values[i], values, like)+like[i]),
    ))
}
